### PR TITLE
Improve apidoc with pydoctor tool 

### PR DESCRIPTION
**Pull request by tisonkun**
Using the [numpy doc style](https://numpydoc.readthedocs.io/en/latest/format.html). This format is also compatible with `pdoc` by specifying `--docformat=numpy`.

To preview locally:

```bash
pip3 install pydoctor
pydoctor --make-html --html-viewsource-base=https://github.com/apache/pulsar-client-python/tree/main --docformat=numpy --theme=readthedocs --html-output=apidoc pulsar
cd apidoc
python3 -m http.server
```

Browse on localhost:8000.

cc @BewareMyPower @merlimat @shibd

### CODE REVIEW

**Comment by tisonkun** [comment]
https://user-images.githubusercontent.com/18818196/200008661-98d2eb9a-6bfd-4114-bb28-7228aca1e51e.mp4

**Comment by tisonkun** [comment]
The major motivation is that the style of pdoc is out of materialized and I suspect it provides worse reading experience.

An alternative is to replace `pdoc` with `pdoc3`:

```
pip3 install pdoc3
pdoc --html -o $(pwd)/apidoc --config "lunr_search={}" pulsar
```

image

But still, we should update several syntax errors. BTW, `pdoc3` is a malicious fork of `pdoc` and IMO Apache basically discourage malicious fork and I'd try to avoid use it as possible.

**Review by merlimat — APPROVED** [review]
👍 👍

**Review by BewareMyPower — COMMENTED** [review]
Could you add a instruction for how to generate the preview into README?

**Comment by tisonkun** [comment]
@BewareMyPower Thanks for your comments. Updated.

File: README.md [hunk 1]
```diff
@@ -110,3 +110,17 @@ Run a single unit test (e.g. `PulsarTest.test_tls_auth`):
...
 ```
+
+## Generate API docs
+
+Pulsar Python Client uses [pydoctor](https://github.com/twisted/pydoctor) to generate API docs. To generate by yourself, run the following command in the root path of this repository:
+
+```bash
+pip3 install pydoctor
+pydoctor --make-html \
+  --html-viewsource-base=https://github.com/apache/pulsar-client-python/tree/<release-version-tag> \
```
Comment (BewareMyPower) [inline README.md:121]:
```suggestion
```

The tag download url is not required for users that want to build the latest docs themselves.

Comment (tisonkun) [reply README.md:121]: This option is to show:

image

... for building the latest docs, you can use `/tree/main` or other refs. This should be inituitive.

Comment (BewareMyPower) [reply README.md:121]: Thanks for your explanation. But I see the source links to an unexpected URL. 

For example, the `class Client: (source)` links to https://github.com/apache/pulsar-client-python/blob/main/pulsar/__init__.py#L344

```python
        _check_type(str, auth_params_string, 'auth_params_string')
```

Do you know the reason? Or does it link correctly in your env?

Comment (tisonkun) [reply README.md:121]: Do you run the command on this patch branch but using the URL prefix of the main branch in upstream?

Comment (tisonkun) [reply README.md:121]: Confirmed: https://github.com/tisonkun/pulsar-client-python/blob/pydoctor/pulsar/__init__.py#L344

**Review by BewareMyPower — APPROVED** [review]

**Pull request merged**